File: backend/src/providers/userProvider.ts
import { BaseSearchRequestType } from '../schemas/search';
import { prisma } from '../lib/prisma.js';
import parseFilterString from '../utils/filterParser.js';
import { UserCreateBodyType, UserUpdateBodyType } from '../schemas/user.js';
import { GuestRegisterBodyType } from '../schemas/validation';

/**
 * Data Access Layer — User
 * All direct Prisma queries for the `users`, `roles`, and `user_roles` tables.
 */
export const userProvider = {
  /** Find all */
  async findAll(request: BaseSearchRequestType) {
    // Use generic parser and provide model-specific hints
    const where = parseFilterString(request.search, {
      allowedFields: ['id', 'name', 'email', 'deleted', 'active', 'phone'],
      fieldTypes: {
        id: 'number',
        name: 'string',
        email: 'string',
        deleted: 'boolean',
        active: 'boolean',
        phoen: 'string',
      },
      defaultSearchFields: ['name', 'email'],
      defaultDeleted: true,
    });

    return prisma.user.findMany({
      where,
      orderBy: { id: request.order },
      take: request.limit + 1,
      ...(request.cursor ? { cursor: { id: request.cursor }, skip: 1 } : {}),
      include: {
        roles: {
          where: { deleted: false },
          include: { role: true },
        },
      },
    });
  },

  /** Find a user by email including their role assignments. */
  async findByEmail(email: string) {
    return prisma.user.findFirst({
      where: { email, deleted: false },
      include: {
        roles: {
          where: { deleted: false },
          include: { role: true },
        },
      },
    });
  },

  /** Find a user by phone including their role assignments. */
  async findByPhone(phone: string) {
    return prisma.user.findFirst({
      where: { phone: phone, deleted: false },
      include: {
        roles: {
          where: { deleted: false },
          include: { role: true },
        },
      },
    });
  },

  /** Find a user by primary key including their role assignments. */
  async findById(id: number) {
    return prisma.user.findFirst({
      where: { id, deleted: false },
      include: {
        roles: {
          where: { deleted: false },
          include: { role: true },
        },
      },
    });
  },

  /** Insert a new user row. Password must already be hashed. */
  async create(data: UserCreateBodyType) {
    return prisma.user.create({ data });
  },

  /** Insert a new user row. Password must already be hashed. */
  async createGuest(data: GuestRegisterBodyType) {
    return prisma.user.create({ data });
  },

  /** Update User */
  async update(data: UserUpdateBodyType, id: number) {
    return prisma.user.update({
      where: { id },
      data,
    });
  },

  /** Update only the avatar (img) field of a user */
  async updateImg(id: number, img: string) {
    return prisma.user.update({
      where: { id },
      data: { img },
    });
  },

  /** Delete a User */
  async delete(id: number) {
    return prisma.user.update({
      where: { id },
      data: { deleted: true },
    });
  },
};

File: backend/app/providers/user_provider.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update
from sqlalchemy.orm import selectinload
from typing import Optional
from app.models.user import User
from app.models.user_role import UserRole
from app.models.role import Role

class UserProvider:
    async def find_by_email(self, db: AsyncSession, email: str) -> Optional[User]:
        result = await db.execute(
            select(User)
            .options(selectinload(User.roles).selectinload(UserRole.role))
            .where(User.email == email, User.deleted == False)
        )
        return result.scalar_one_or_none()

    async def find_by_phone(self, db: AsyncSession, phone: str) -> Optional[User]:
        result = await db.execute(
            select(User).where(User.phone == phone, User.deleted == False)
        )
        return result.scalar_one_or_none()

    async def find_by_id(self, db: AsyncSession, user_id: int) -> Optional[User]:
        result = await db.execute(
            select(User)
            .options(selectinload(User.roles).selectinload(UserRole.role))
            .where(User.id == user_id)
        )
        return result.scalar_one_or_none()

    async def create(self, db: AsyncSession, **data) -> User:
        user = User(**data)
        db.add(user)
        await db.commit()
        await db.refresh(user)
        return user

    async def create_guest(self, db: AsyncSession, **data) -> User:
        user = User(**data)
        db.add(user)
        await db.commit()
        await db.refresh(user)
        return user

    async def find_all(self, db: AsyncSession, limit: int = 10, cursor: Optional[int] = None, order: str = "desc", search: Optional[str] = None):
        from sqlalchemy import asc, desc
        stmt = select(User).options(selectinload(User.roles).selectinload(UserRole.role)).where(User.deleted == False)
        if search:
            stmt = stmt.where(User.name.ilike(f"%{search}%") | User.email.ilike(f"%{search}%"))
        if cursor:
            if order == "desc":
                stmt = stmt.where(User.id < cursor)
            else:
                stmt = stmt.where(User.id > cursor)
        stmt = stmt.order_by(desc(User.id) if order == "desc" else asc(User.id))
        stmt = stmt.limit(limit + 1)
        result = await db.execute(stmt)
        return result.scalars().all()

    async def update(self, db: AsyncSession, user_id: int, **data) -> Optional[User]:
        await db.execute(update(User).where(User.id == user_id).values(**data))
        await db.commit()
        return await self.find_by_id(db, user_id)

    async def update_avatar(self, db: AsyncSession, user_id: int, img: str) -> Optional[User]:
        await db.execute(update(User).where(User.id == user_id).values(img=img))
        await db.commit()
        return await self.find_by_id(db, user_id)

    async def soft_delete(self, db: AsyncSession, user_id: int) -> Optional[User]:
        await db.execute(update(User).where(User.id == user_id).values(deleted=True))
        await db.commit()
        return await self.find_by_id(db, user_id)

user_provider = UserProvider()
